include/pybind11/pybind11.h

```diff
@@ -1040,12 +1041,36 @@ class generic_type : public object {
 
         auto &internals = get_internals();
         auto tindex = std::type_index(*rec.type);
+        auto module_local = rec.module_local;
         tinfo->direct_conversions = &internals.direct_conversions[tindex];
-        if (rec.module_local)
+        if (module_local)
             registered_local_types_cpp()[tindex] = tinfo;
         else
             internals.registered_types_cpp[tindex] = tinfo;
-        internals.registered_types_py[(PyTypeObject *) m_ptr] = { tinfo };
+        internals.registered_types_py[type] = { tinfo };
+
+        // Clean up our internals after the Python type object gets garbage collected
+        weakref(m_ptr, cpp_function([type, tindex, module_local](handle wr) {
+            auto &internals = get_internals();
+            internals.direct_conversions.erase(tindex);
+            if (module_local)
+                registered_local_types_cpp().erase(tindex);
+            else
+                internals.registered_types_cpp.erase(tindex);
+            internals.registered_types_py.erase(type);
+            // C++20: std::erase_if(internals.inactive_override_cache,
```

Comment:
nit While this is cool, I think it's a bit o' clutter. Would you be able to delete this comment?

(You could make a permalink, and then reference it here? e.g.
```
// TODO: Use C++20 code (see commit X on PR Y).
```

Reply:
Honestly, I don't mind it (and I hate commented out code). It shows the intended logic in a more concise way especially readable to someone used to C++20. But commented out code becomes "stale", and is untested, so I'd also be okay to see it go.

Reply:
As you wish. I was just frustrated that C++11 didn't have `erase_if` (and `remove_if` is not really applicable to `unordered_set`, I think).

I don't think the permalink would make it so much better, then. I can just remove it and put just something like `// Actually just `std::erase_if`, but only available in C++20`. By the time we move to C++20, I'm sure it'll be easy enough to clean up?

Reply:
Forgot about this. Let me still quickly fix this!